src/main/scala/com/fulcrumgenomics/vcf/FixVcfPhaseSet.scala

```diff
@@ -0,0 +1,250 @@
+      .find(value => value.id == "PS" && value.count == VcfCount.Fixed(1) && value.kind == VcfFieldType.Integer)
+      .getOrElse {
+          VcfFormatHeader(
+          id          = "PS",
+          count       = VcfCount.Fixed(1),
+          kind        = VcfFieldType.Integer,
+          description = "Phasing set (typically the position of the first variant in the set)"
+        )
+      }
+
+    // In some cases, input VCFs may have heeder lines (ex. PEDIGREE) with no ID, which is required
+    val others = if (removeNoIdHeaderLines) header.others.filterNot(_.id == null) else header.others
+
+    // Put it all together
+    header.copy(
+      formats = header.formats.filter(_.id != "PS") ++ original.iterator.toSeq :+ newPhaseSet,
+      others  = others
+    )
+  }
+}
+
+object VcfPhaseSetUpdater {
+  /** Base traits for the result of  [[VcfPhaseSetUpdater.updateGenotype()]] */
+  sealed trait Result extends EnumEntry {
+    def genotype: Genotype
+    def description: String
+  }
+  object Result extends FgBioEnum[Result] {
+    case class Valid                      (genotype: Genotype) extends Result { val description: String = "were valid" }
+    case class UpdatedPhaseSetValue       (genotype: Genotype) extends Result { val description: String = "had phase set value updated only" }
+    case class UpdatedPhaseAndValue       (genotype: Genotype) extends Result { val description: String = "were set to phased and had phase set value updated" }
+    case class UpdatedPhaseOnly           (genotype: Genotype) extends Result { val description: String = "were set to phased" }
+    case class PhasedMissingPhaseSetValue (genotype: Genotype) extends Result { val description: String = "were phased but missing a phase set value" }
+    case class NotPhasedWithPhaseSetValue (genotype: Genotype) extends Result { val description: String = "were not phased but had a phase set value" }
+    case class NotPhasedNoPhaseSetValue   (genotype: Genotype) extends Result { val description: String = "were not phased and no phase set value" }
+    override def values: immutable.IndexedSeq[Result] = findValues
+  }
+}
+
+class VcfPhaseSetUpdater(header: VcfHeader, keepOriginal: Boolean, phaseGenotypesWithPhaseSet: Boolean) extends LazyLogging {
+  import VcfPhaseSetUpdater._
+  import VcfPhaseSetUpdater.Result._
+
+  private val phaseSetToPositionBySample: Map[String, mutable.HashMap[String, Int]] = {
+    header.samples.map { sample => (sample, scala.collection.mutable.HashMap[String, Int]()) }.toMap
+  }
+  private var lastChrom: String = ""
+  private var variants: Long    = 0
+  private var genotypes: Long   = 0
+  val descriptionCounter: SimpleCounter[String] = new SimpleCounter()
+
+  def update(variant: Variant): Variant = {
+    // Phase sets cannot span contigs!  So empty our mappings if we get to a new one
+    if (variant.chrom != lastChrom) {
+      this.phaseSetToPositionBySample.values.foreach(_.clear())
+      this.lastChrom = variant.chrom
+    }
```

Comment:
Just looking at this, I'm guessing the solution below requires the VCF to be sorted?  Given that it's borked up in other ways do you want to check/assert that as you're iterating over the variants?  Maybe the chromosome ordering doesn't matter, but that you see each chromosome in a contiguous block of variants and the variants have rising positions?